---
title: "GWAS Circos Plot"
author:
  - "**[Editor]** [Benben Miao](https://github.com/benben-miao)"
---

The visualization of Genome-Wide Association Study (GWAS) results mainly includes SNP circular plots displayed by chromosome positions, SNP density plots, Manhattan plots for significance screening, QQ plots comparing the distribution of observed p-values with expected p-values, etc., which are used to screen candidate variant genes at the genome-wide level.

## Example

![](../images/Omics/GwasSnpPlot_demo.png){fig-alt="GwasSnpPlot DEMO" fig-align="center" width="60%"}

## Setup

-   System Requirements: Cross-platform (Linux/MacOS/Windows)

-   Programming language: R

-   Dependent packages: `MetaNet`; `pcutils`; `igraph`; `dplyr`

```{r packages setup, message=FALSE, warning=FALSE, output=FALSE}
# Install packages
if (!requireNamespace("CMplot", quietly = TRUE)) {
  install.packages("CMplot")
}

# Library packages
library(CMplot)
```

```{r}
sessioninfo::session_info("attached")
```

## Data Preparation

The sample data includes 60K liquid chip raw variation information data of Pig.

```{r load data, message=FALSE, warning=FALSE}
# Example data
data(pig60K)
data <- pig60K

# Data preview
head(data, 5)
```

## Visualization

### 1. SNP screening genome circular map

Genome circular maps can display more chromosomes and traits under the same proportion, and have a great advantage when comparing multiple traits.

```{r fig1GwasSnpCircos}
#| label: fig-1GwasSnpCircos
#| fig-cap: "SNP screening genome circular map"
#| out.width: "95%"
#| warning: false

# SNP screening genome circular map
p <- CMplot(
  data,
  type = "p",
  plot.type = "c",
  chr.labels = paste("Chr", c(1:18, "X", "Y"), sep = ""),
  r = 8,
  cir.axis = TRUE,
  outward = TRUE,
  cir.axis.col = "black",
  cir.chr.h = 2,
  chr.den.col = "black",
  file.output = FALSE,
  verbose = FALSE,
  mar = c(0,0,0,0)
)

p
```

### 2. SNP Chromosome Density Map

The SNP chromosome density map shows the distribution of SNP density on each chromosome.

```{r fig2GwasSnpDensity}
#| label: fig-2GwasSnpDensity
#| fig-cap: "SNP Chromosome Density Map"
#| out.width: "95%"
#| warning: false
#| message: false

# SNP Chromosome Density Map
p <- capture.output(
  CMplot(
    data,
    plot.type = "d",
    bin.size = 1e6,
    chr.den.col = c("darkgreen", "yellow", "red"),
    main = "Density",
    file.output = FALSE,
    verbose = FALSE
  )
)

print(p[1])
```

### 3. Manhattan plot for SNP screening

The Manhattan plot for SNP screening is convenient for displaying the significance of SNPs at 5% or 1%, and it also helps identify the concentration effect of key chromosomes.

```{r fig3GwasSnpManhattan}
#| label: fig-3GwasSnpManhattan
#| fig-cap: "Manhattan plot for SNP screening"
#| out.width: "95%"
#| warning: false
#| message: false

# Manhattan plot for SNP screening
p <- CMplot(
  data,
  type = "p",
  plot.type = "m",
  LOG10 = TRUE,
  threshold = NULL,
  file.output = FALSE,
  verbose = FALSE,
  chr.labels.angle = 45
)

p
```

### 4. QQ plot based on SNP

The core function of the QQ plot based on SNPs is to test whether the P-value distribution of SNPs in GWAS analysis conforms to the "expected distribution under the no-association hypothesis", so as to judge the reliability of the analysis results.

```{r fig4GwasSnpQQ}
#| label: fig-4GwasSnpQQ
#| fig-cap: "QQ plot based on SNP"
#| out.width: "95%"
#| warning: false
#| message: false

# QQ plot based on SNP
p <- CMplot(
  data,
  plot.type = "q",
  box = FALSE,
  conf.int = TRUE,
  conf.int.col = NULL,
  threshold.col = "red",
  threshold.lty = 2,
  file.output = FALSE,
  verbose = FALSE
)

p
```