---
title: "Multiple Sequences Alignment"
author:
  - "**[Editor]** [Benben Miao](https://github.com/benben-miao)"
---

Multiple Sequence Alignment (MSA) is a fundamental and crucial technique in bioinformatics. It is used to align three or more biological sequences (DNA, RNA, or proteins) based on their evolutionary or structural similarities, so that homologous sites (i.e., sites derived from a common ancestor) are aligned as much as possible.

## Example

![](../images/Omics/MultiSeqsAlignment.png){fig-alt="MultiSeqsAlignment" fig-align="center" width="60%"}

## Setup

-   System Requirements: Cross-platform (Linux/MacOS/Windows)

-   Programming language: R

-   Dependent packages: `ggmsa`

```{r packages setup, message=FALSE, warning=FALSE, output=FALSE}
# Install packages
if (!requireNamespace("ggmsa", quietly = TRUE)) {
  install.packages("ggmsa")
}

# Library packages
library(ggmsa)
```

```{r}
sessioninfo::session_info("attached")
```

## Data Preparation

DNA/RNA/amino acid sequence data is usually stored in FASTA format, where each sequence consists of a description line (starting with ">") followed by sequence lines.

```{r load data, message=FALSE, warning=FALSE}
# Example data
protein_fasta <- system.file("extdata", "sample.fasta", package = "ggmsa")

# Data preview
seqs <- readLines(protein_fasta)
head(seqs)
```

## Visualization

### 1. Multiple sequence alignment of proteins

The visualization of multiple protein sequence alignment shows the alignment of multiple protein sequences in specific regions, helps identify conserved regions and mutation sites, and provides a professional color scheme for amino acid sequences.

```{r fig1MultiSeqsAlignmentProtein}
#| label: fig-1MultiSeqsAlignmentProtein
#| fig-cap: "Multiple sequence alignment of proteins"
#| out.width: "95%"
#| warning: false

# Multiple sequence alignment of proteins
p <- ggmsa(
  protein_fasta,
  start = 300,
  end = 330,
  font = "DroidSansMono",
  color = "Chemistry_AA",
  char_width = 0.5,
  seq_name = TRUE,
  consensus_views = FALSE
)

p
```

### 2. Multiple sequence alignment and statistics

The statistical chart of multiple sequence alignment results shows the conservation and variation of multiple sequences within the alignment region, helping to identify key sites and regions.

```{r fig1MultiSeqsAlignmentStats}
#| label: fig-1MultiSeqsAlignmentStats
#| fig-cap: "Multiple sequence alignment and statistics"
#| out.width: "95%"
#| warning: false

# Multiple sequence alignment and statistics
p <- ggmsa(
  protein_fasta,
  start = 300,
  end = 320,
  font = "DroidSansMono",
  color = "Chemistry_AA",
  char_width = 0.5,
  seq_name = TRUE,
  consensus_views = FALSE
) +
  geom_msaBar()

p
```

### 3. Multiple sequence alignment and Logo

While performing multiple sequence alignment and displaying the alignment consensus results with a Logo, the bases or amino acids of multiple sequences are sorted and displayed in the form of a Logo.

```{r fig1MultiSeqsAlignmentLogo}
#| label: fig-1MultiSeqsAlignmentLogo
#| fig-cap: "Multiple sequence alignment and Logo"
#| out.width: "95%"
#| warning: false

# Multiple sequence alignment and Logo
p <- ggmsa(
  protein_fasta,
  start = 300,
  end = 330,
  font = "DroidSansMono",
  color = "Chemistry_AA",
  char_width = 0.5,
  seq_name = TRUE,
  consensus_views = FALSE
) +
  geom_seqlogo()

p
```